File: c#/Sort/TreeSort.cs
using System.Collections;


namespace Sort
{
    public class Tree
    {
        private Tree _left;
        private Tree _right;
        private int _key;

        public Tree Left
        {
            get { return _left; }
            set
            {
                if(null == _left)
                    _left = value;
                else
                    _left.Insert(value);
                
            }
        }

        public Tree Right
        {
            get { return _right; }
            set
            {
                if (null == _right)
                    _right = value;
                else
                    _right.Insert(value);
            }
        }



        public Tree(int key)
        {
            _key = key;
        }

        public void Insert(Tree tree)
        {
            if(tree._key < _key)
                Left = tree;
            else
                Right = tree;
        }

        public ArrayList ToArrayList()
        {
            var result = new ArrayList();
            if (null != Left)
                result.AddRange( Left.ToArrayList() );
            result.Add(_key);
            if (null != Right)
                result.AddRange(Right.ToArrayList());

            return result;
        }
       
    }

    public class TreeSort : ISort
    {
        public ArrayList Sort(ArrayList inArray)
        {
            if (inArray.Count < 1)
                return inArray;

            var tree = new Tree((int)inArray[0]);
            for (int i = 1; i < inArray.Count; i++ )
                tree.Insert(new Tree((int)inArray[i]));

            return tree.ToArrayList();
        }
    }
}

File: python/src/sort/tree_sort.py
def tree_sort(in_array):
    if len(in_array) < 2:
        return in_array

    tree = Tree(in_array[0])
    for i in range(1,len(in_array)):
        tree.insert(Tree(in_array[i]))

    return tree.to_array()

class Tree():
    def __init__(self, key):
        self.__key = key
        self.__left = None
        self.__right = None


    def insert(self, tree):

        if self.key > tree.key:
            self.left = tree
        else:
            self.right = tree

    def to_array(self):
        result=[]
        if self.left != None:
            result += self.left.to_array()
        result += [self.__key]
        if self.right != None:
            result += self.right.to_array()

        return result

    @property
    def key(self):
        return self.__key

    @property
    def left(self):
        return self.__left

    @left.setter
    def left(self, tree):
        if self.__left == None:
            self.__left = tree
        else:
            self.__left.insert(tree)

    @property
    def right(self):
        return self.__right

    @right.setter
    def right(self, tree):
        if self.__right == None:
            self.__right = tree
        else:
            self.__right.insert(tree)